### PR TITLE
#1106 Add toggle control to use in Modeler modeling node overrides

### PR DESCRIPTION
**Pull request by nabilIbm**
Fixes : #1106 

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Review by matthoward366 — CHANGES_REQUESTED** [review]

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		this.id = ControlUtils.getControlId(this.props.propertyId);
+	}
+
+	handleChange(value) {
+		this.props.controller.updatePropertyValue(this.props.propertyId, value);
+	}
+
+	render() {
+		const label = this.props.control.label ? this.props.control.label.text : "";
+		const tooltipId = uuid4() + "-tooltip-" + this.props.control.name;
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx]: Took textarea as reference.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		return (
+			<div className={classNames("properties-toggle", { "hide": this.props.state === STATES.HIDDEN }, this.props.messageInfo ? this.props.messageInfo.type : null)}
+				data-id={ControlUtils.getDataId(this.props.propertyId)}
+			>
+				<Toggle
+					disabled={this.props.state === STATES.DISABLED}
+					id={this.id}
+					size="sm"
+					labelText={toggleLabel}
+					onToggle={this.handleChange.bind(this)}
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx]: added toggled with value being passed in.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+.properties-toggle {
+	display: flex;
+	align-items: flex-end;
+	max-width: 100%;
+	width: 50%;
+	.bx--toggle-input__label {
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss]: removed it.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	.tooltip-container {
+		margin-left: $spacing-03;
+		margin-bottom: $spacing-02;
+		cursor: pointer;
+		.tooltip-trigger {
+			display: flex;
+		}
+	}
+}
+.properties-ft-control-container {
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss]: removed it.

File: canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+			/>
+		);
+		expect(wrapper.prop("control")).to.equal(control);
+		expect(wrapper.prop("propertyId")).to.equal(propertyId);
+		expect(wrapper.prop("controller")).to.equal(controller);
+	});
+
+});
+
+describe("toggle classnames appear correctly", () => {
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js:76]: OK

**Comment by nmgokhale** [comment]
In "Conditions" category, there's no error or warning displayed when I change the toggle.  
There's a toggle added for "hide" and "Disable" - it just shows label "On". Its unclear what these toggles are doing. They should be checkboxes. 
Screen Shot 2022-07-28 at 1 18 50 PM

Following table never shows any warnings or errors. Also description says Type error and type warn. This doesn't apply to the toggle table. 
Screen Shot 2022-07-28 at 1 22 41 PM

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	render() {
+		const value = this.props.value ? this.props.value : "";
+		const toggleControl = (<Toggle
+			id={this.id}
+			size="sm"
+			disabled={this.props.state === STATES.DISABLED}
+			toggled={Boolean(this.props.value)}
+			labelA={this.props.toggleofflabel}
+			labelB={this.props.toggleonlabel}
+			onToggle={this.handleChange.bind(this)}
```
Comment (nmgokhale) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:49]: Please define [labelText](https://v10-react.carbondesignsystem.com/?path=/docs/components-toggle--default) here for accessibility -

`labelText - Provide the text that will be read by a screen reader when visiting this control`

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+      {
+        "parameter_ref": "string",
+        "label": {
+          "default": "string"
+        },
+        "control": "toggle",
+        "description": {
+          "default": "toggle with sting value"
+        },
+        "class_name": "string-toggle-control-class"
```
Comment (nmgokhale) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:168]: Please add `toggleonlabel` and `toggleofflabel` in paramDef for testing.

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 8]
```diff
@@ -0,0 +1,507 @@
+{
+  "titleDefinition": {
+    "title": "toggle Fields"
+  },
+  "current_parameters": {
+    "string": "toggle",
```
Comment (nmgokhale) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:6]: You're passing this value in paramDef. But I don't see this anywhere on UI. 
Since toggle is a boolean control, what does it have string value?

Screen Shot 2022-07-28 at 1 46 57 PM

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+		this.props.controller.updatePropertyValue(this.props.propertyId, value);
+	}
+
+	render() {
+		const value = this.props.value ? this.props.value : "";
+		const toggleControl = (<Toggle
+			id={this.id}
+			size="sm"
+			disabled={this.props.state === STATES.DISABLED}
+			toggled={Boolean(this.props.value)}
```
Comment (nmgokhale) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:46]: When `this.props.value` is a string. `Boolean(this.props.value)` returns true. This is incorrect. 
`this.props.value` should always be boolean. And `toggled={this.props.value}`

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+ToggleControl.propTypes = {
+	control: PropTypes.object.isRequired,
+	propertyId: PropTypes.object.isRequired,
+	controller: PropTypes.object.isRequired,
+	controlItem: PropTypes.element,
+	tableControl: PropTypes.bool,
+	state: PropTypes.string, // pass in by redux
+	value: PropTypes.string,
```
Comment (nmgokhale) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:93]: This should be boolean

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+			disabled={this.props.state === STATES.DISABLED}
+			toggled={Boolean(this.props.value)}
+			labelA={this.props.toggleofflabel}
+			labelB={this.props.toggleonlabel}
+			onToggle={this.handleChange.bind(this)}
+		/>);
+
+
+		let display = toggleControl;
+		if (this.props.tableControl) {
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:54]: This shouldn't be needed.  I don't see a reason to add a tooltip in a table cell for a toggle.  This was used in textfield since there could be an overflow.

File: canvas_modules/common-canvas/src/common-properties/form/ControlInfo.js [hunk 12]
```diff
@@ -71,6 +71,12 @@ export class Control {
 		if (settings.charLimit) {
 			this.charLimit = settings.charLimit;
 		}
+		if (settings.toggleofflabel) {
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/form/ControlInfo.js:74]: I don't think this make sense.  These need to be translatable strings and this wouldn't provide that.

**Review by matthoward366 — CHANGES_REQUESTED** [review]

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx (hunk 10 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:68]: Adjusted.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx (hunk 6 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:45]: Adjusted it.

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json (hunk 7 above)
Comment (nabilIbm) [reply canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:148]: image

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx (hunk 11 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx]: @nmgokhale fixed the error issues 
image

File: canvas_modules/common-canvas/src/common-properties/form/ControlInfo.js (hunk 12 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/form/ControlInfo.js]: I think since we're reading these strings from parameter definition, these will be translatable strings. Same way label and description are translatable as it's coming from paramDef.

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 13]
```diff
@@ -0,0 +1,484 @@
...
+          "Babbage"
+        ]
+      ],
+      []
+    ]
+  },
+  "parameters": [
+    {
+      "id": "string",
+      "type": "string"
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:53]: Since toggles are boolean we shouldn't be defining everything as `string`.  These should be boolean.  The names are also wrong.  I would probably just name them `toggle_...`

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+			disabled={this.props.state === STATES.DISABLED}
+			toggled={Boolean(this.props.value)}
+			labelA={this.props.toggleofflabel}
+			labelB={this.props.toggleonlabel}
+			onToggle={this.handleChange.bind(this)}
+			labelText={this.props.controlItem}
+		/>);
+
+
+		const className = classNames("properties-toggle", { "hide": this.props.state === STATES.HIDDEN }, this.props.messageInfo ? this.props.messageInfo.type : null);
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:50]: Is `this.props.messageInfo ? this.props.messageInfo.type : null` used for anything?  I don't think it is.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx (hunk 9 above)
Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:42]: This wasn't fixed.

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json (hunk 7 above)
Comment (matthoward366) [reply canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:148]: Was this pushed. This wouldn't be the correct way of handling labels since they need to be translatable.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+.properties-toggle {
+	.bx--toggle-input__label .bx--toggle__switch {
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss:19]: Why is this needed?  We only use carbon's classes when needed.

File: canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+				controller={controller}
+				propertyId={propertyId}
+			/>
+		);
+		expect(wrapper.prop("control")).to.equal(control);
+		expect(wrapper.prop("propertyId")).to.equal(propertyId);
+		expect(wrapper.prop("controller")).to.equal(controller);
+	});
+
+	it("toggle should render correctly", () => {
```
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js:58]: There aren't any tests to validating toggling the control actually works.

File: canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+		const wrapper = mount(
+			<Toggle
+				store={controller.getStore()}
+				control={control}
+				controller={controller}
+				propertyId={propertyId}
+			/>
+		);
+		const toggleWrapper2 = wrapper.find("span.bx--toggle__switch");
```
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js:67]: These don't really test anything.  It's best to stay away from carbon specific class when not needed.  We don't need to text that carbon shows the correct classes only that the control gets rendered and we can change the value.

Testing custom labels would also be imported.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+		this.props.controller.updatePropertyValue(this.props.propertyId, value);
+	}
+
+	render() {
+		const toggleControl = (<Toggle
+			id={this.id}
+			size="sm"
+			disabled={this.props.state === STATES.DISABLED}
+			toggled={Boolean(this.props.value)}
+			labelA={this.props.toggleofflabel}
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:43]: I'm not sure how this works.  These need to come from the translatable resources.

**Review by matthoward366 — CHANGES_REQUESTED** [review]

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx (hunk 18 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx]: Fixed them using` this.props.controller.getResource(resourceKey)`

File: canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+		const toggle = wrapper.find("div.properties-toggle");
+
+		const image = toggle.find("svg");
+		expect(image).to.have.length(1);
+		image.simulate("click");
+
+		const toggleLableOn = toggle.find("span.bx--toggle__text--on");
+		const toggleLableOff = toggle.find("span.bx--toggle__text--off");
+
+		expect(toggleLableOn.text()).to.equal(control.values[0]);
```
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js:90]: This isn't testing that the value changed.  This is just validating the labels as far as I can tell.  You can get the updated value from the propertiesController.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+ToggleControl.propTypes = {
+	control: PropTypes.object.isRequired,
+	propertyId: PropTypes.object.isRequired,
+	controller: PropTypes.object.isRequired,
+	controlItem: PropTypes.element,
+	tableControl: PropTypes.bool,
+	state: PropTypes.string, // pass in by redux
+	value: PropTypes.bool,
+	toggleofflabel: PropTypes.string,
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:80]: I don't think these are used.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+ToggleControl.propTypes = {
+	control: PropTypes.object.isRequired,
+	propertyId: PropTypes.object.isRequired,
+	controller: PropTypes.object.isRequired,
+	controlItem: PropTypes.element,
+	tableControl: PropTypes.bool,
+	state: PropTypes.string, // pass in by redux
+	value: PropTypes.bool,
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:79]: Can you add this comment `// pass in by redux`

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+
+		);
+	}
+}
+
+ToggleControl.propTypes = {
+	control: PropTypes.object.isRequired,
+	propertyId: PropTypes.object.isRequired,
+	controller: PropTypes.object.isRequired,
+	controlItem: PropTypes.element,
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:76]: Is this used?

Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.jsx:76]: Im using it to fetch `labelText` content
image

File: canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+				control={control}
+				controller={controller}
+				propertyId={propertyId}
+			/>
+		);
+		const toggleWrapper = wrapper.find("div.properties-toggle");
+		const image = toggleWrapper.find("svg");
+		expect(image).to.have.length(1);
+		image.simulate("click");
+		expect(controller.getPropertyValue(propertyId)).to.equal(true);
```
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js:93]: I don't think this correct.  The initial value is `true`.  On a click shouldn't this be false?  To make it more obvious in the test you can check the value before and after to make sure it changes.

File: canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json [hunk 24]
```diff
@@ -0,0 +1,429 @@
...
+    },
+    {
+      "id": "toggle_hidden",
+      "type": "boolean",
+      "required": true
+    }
+  ],
+  "complex_types": [
+    {
+      "id": "togglees",
```
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json:77]: spelling

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 25]
```diff
@@ -0,0 +1,433 @@
+{
+  "titleDefinition": {
+    "title": "toggle Fields"
+  },
+  "current_parameters": {
+    "toggleـstring_null": true,
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:6]: Can you cleanup the naming here?  These are also wrong.  For example this one should be "toggle_null": null.  For undefined there wouldn't be a current parameter.

Also remove `string` from all the names since it doesn't make sense.

**Comment by matthoward366** [comment]
Screen Shot 2022-08-25 at 3 33 13 PM

In tables can the toggle be centered?

**Comment by matthoward366** [comment]
Once center you'll also need to fix the warning/error icon so they don't overlap.  
Screen Shot 2022-08-25 at 3 34 54 PM

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json (hunk 25 above)
Comment (nabilIbm) [reply canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json]: Updated `toggle_null` & removed string

File: canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json (hunk 24 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json:77]: Adjuted it.

**Comment by nabilIbm** [comment]
image

File: canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js (hunk 23 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js]: I don't think clicking something is changing the value of toggle ,In toggle.jsx file  toggled={this.props.value} And value: ownProps.controller.getPropertyValue(ownProps.propertyId), 
So whatever boolean "value" is been set in properties controller - "toggled" property is getting that value.
So IInitially I set 
```
controller.setPropertyValues(
	{ toggle: true }
);
``` Then verified the prop `toggled`
<img width="654" alt="image" src="https://user-images.githubusercontent.com/89609176/187263102-166feafb-758c-499b-b3b4-aa0f16fa639e.png">

File: canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+		controller.setErrorMessages({});
+		controller.setControlStates({});
+		controller.setPropertyValues(
+			{ toggle: true }
+		);
+	});
+
+
+	it("toggle should set correct value", () => {
+		// In toggle.jsx we're setting toggled={this.props.value}
```
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/common-properties/controls/toggle-test.js:51]: This test doesn't really do anything.  You need to test where clicking on the toggle changes the value.

File: canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json [hunk 27]
```diff
@@ -0,0 +1,429 @@
...
+        },
+        "control": "toggle",
+        "description": {
+          "default": "toggle that is hidden"
+        }
+      }
+    ],
+    "complex_type_info": [
+      {
+        "complex_type_ref": "togglees",
```
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json:222]: spelling

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+.properties-toggle {
+	margin-bottom: 14px;
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss:18]: Why is this 14px? If we need a margin this should be a token.

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 29]
```diff
@@ -0,0 +1,432 @@
...
+        ],
+        [
+          false
+        ]
+      ]
+    ]
+  },
+  "parameters": [
+    {
+      "id": "toggle_table__default",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:50]: why does this have table in the name?  This isn't a table.

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 30]
```diff
@@ -0,0 +1,432 @@
...
+      "type": "boolean"
+    },
+    {
+      "id": "toggle_table__undefined",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:54]: same as above

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 31]
```diff
@@ -0,0 +1,432 @@
...
+      "type": "boolean"
+    },
+    {
+      "id": "toggleـnull",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:58]: Can you change this to toggle_null to be consistent?

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 32]
```diff
@@ -0,0 +1,432 @@
...
+      "type": "boolean"
+    },
+    {
+      "id": "toggle_table__error",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:62]: This isn't a table

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 33]
```diff
@@ -0,0 +1,432 @@
...
+      "type": "boolean"
+    },
+    {
+      "id": "toggleـwarning",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:66]: rename

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 34]
```diff
@@ -0,0 +1,432 @@
...
+      "type": "boolean"
+    },
+    {
+      "id": "toggle_tablehide",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:70]: not a table

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 35]
```diff
@@ -0,0 +1,432 @@
...
+      "type": "boolean"
+    },
+    {
+      "id": "toggle_table__hidden",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:74]: not a table

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 36]
```diff
@@ -0,0 +1,432 @@
...
+      "type": "boolean"
+    },
+    {
+      "id": "disable",
+      "type": "boolean"
+    },
+    {
+      "id": "toggle_table__disabled",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:82]: not a table

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 37]
```diff
@@ -0,0 +1,432 @@
...
+        {
+          "id": "toggle",
+          "type": "boolean"
+        },
+        {
+          "id": "toggle2",
+          "type": "boolean"
+        },
+        {
+          "id": "toggle_tableon_panel",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:107]: probably should just be `toggle_on_panel`

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 38]
```diff
@@ -0,0 +1,432 @@
...
+          "type": "boolean"
+        },
+        {
+          "id": "toggle_tablesub_panel",
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:111]: toggle_sub_panel

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 39]
```diff
@@ -0,0 +1,432 @@
...
+          "type": "boolean"
+        }
+      ]
+    }
+  ],
+  "uihints": {
+    "id": "toggle",
+    "icon": "images/default.svg",
+    "label": {
+      "default": "_ Fields"
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:121]: Is this the correct label?

File: canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json [hunk 40]
```diff
@@ -0,0 +1,432 @@
+{
+  "titleDefinition": {
+    "title": "toggle Fields"
+  },
+  "current_parameters": {
+    "toggleـnull": null,
+    "toggle_table__error": false,
+    "toggleـwarning":false,
+    "toggle_table__undefined":false,
```
Comment (matthoward366) [inline canvas_modules/harness/test_resources/parameterDefs/toggle_paramDef.json:9]: This isn't undefined.  You have the value set to `false`

**Review by matthoward366 — CHANGES_REQUESTED** [review]

File: canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json (hunk 27 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json:222]: fixed.

File: canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss (hunk 28 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss]: used token.

Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss]: What is this for?

Comment (nabilIbm) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss]: Used it to center toggles in tables , isnt the right approach?

without token
image
with token
image

Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss]: No it adds too much space to the toggle when not in tables.  

When in a table you should override `bx--toggle__switch` and set the margin to 0.  This is only for tables though.  Also make sure to add a comment.
Screen Shot 2022-08-30 at 1 03 55 PM

Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/controls/toggle/toggle.scss]: Looks better.  Can you use this
```
.properties-table-cell-control { // overrides when in a table cell
	.properties-toggle {
		.bx--toggle__switch { // override carbon when inside a table
			margin: $spacing-02 0;
		}
	}
}
```

File: canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json (hunk 24 above)
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json:77]: spelling

File: canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json (hunk 27 above)
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json:222]: spelling

**Review by matthoward366 — COMMENTED** [review]
A few minor comment.  You'll also need to update https://github.com/elyra-ai/pipeline-schemas/blob/main/common-pipeline/operators/uihints-v3-schema.json#L256 to add toggle.

File: canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json (hunk 24 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json:77]: image

Comment (matthoward366) [reply canvas_modules/common-canvas/__tests__/test_resources/paramDefs/toggle_paramDef.json:77]: This is wrong in the PR yet.  Maybe you fix it in the harness and didn't sync it in the test case.

File: canvas_modules/common-canvas/locales/common-properties/locales/en.json [hunk 41]
```diff
@@ -67,6 +67,8 @@
   "expression.all.functions": "All Functions",
   "expression.min.label": "Min",
   "expression.max.label": "Max",
+	"MESSAGE_KEYS.TOGGLE_ON_LABEL": "toggle.on.label",
```
[the remark this answers is not in the document]
Comment (matthoward366) [reply canvas_modules/common-canvas/locales/common-properties/locales/en.json:70]: Also, fix the indent.

[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/locales/common-properties/locales/en.json]: image

File: canvas_modules/common-canvas/locales/common-properties/locales/en.json [hunk 42]
```diff
@@ -67,6 +67,8 @@
   "expression.all.functions": "All Functions",
   "expression.min.label": "Min",
   "expression.max.label": "Max",
+  "toggle.on.label": "on",
```
Comment (matthoward366) [inline canvas_modules/common-canvas/locales/common-properties/locales/en.json:70]: Should be On and Off.

Comment (nabilIbm) [reply canvas_modules/common-canvas/locales/common-properties/locales/en.json]: image

**Review by matthoward366 — APPROVED** [review]
Looks good
